### PR TITLE
Add param to allow customizing boosting of fields searched

### PR DESCRIPTION
**Pull request by xyu**
#### Changes proposed in this Pull Request:

This adds a param which can for now only be modified via filters which provide more control over how much importance we place on each field that is searched. Currently the following field boost values can be adjusted via this new `excess_boost` param:

* `title`
* `tag.name`
* `category.name`
* `author_login`
* `author`

The following fields are searched however they can not be boosted when terms are found in them:

* `content`
* `excerpt`

The boosting given is in excess of preconfigured boost and are multiplied to it thus a positive number must be given.

Caution: We may at some point adjust the default baseline boosting to provide better results which may amplify any manual fine tuning done with this filter.

#### Testing instructions:

This filter can be tested by adding something like the following then checking if results with the given query in tags is ranked higher than before the change.

```
/**
 * Give tags more weight in searches
 *
 * @param array    $es_wp_query_args
 * @param WP_Query $query
 *
 * @return array   The modified $es_wp_query_args
 */
function myexample_weigh_tags( array $es_wp_query_args, WP_Query $query ) {
  if ( !is_array( $es_wp_query_args['excess_boost'] ) ) {
    $es_wp_query_args['excess_boost'] = array();
  }

  // Add boost for tags
  $es_wp_query_args['excess_boost']['tag.name'] = 2.5;

  return $es_wp_query_args;
}
add_filter( 'jetpack_search_es_wp_query_args', 'myexample_weigh_tags', 10, 2 ); 
```

#### Proposed changelog entry for your changes:

* Jetpack Search: Adds an advanced `excess_boost` param which can be adjusted with filters to fine tune query scoring.

### CODE REVIEW

**Comment by brbrr** [comment]
Passing-by comment: using both "In Progress" & "Needs Review" labels at the same time is confusing. Is it ready for review?

**Comment by xyu** [comment]
@bperson great idea on making the code cleaner and DRYer, can I get another review after the latest changes?

> Oh should we also make best_fields / most_fields parametrizable?

I thought about this but decided against it because I figured if we open it up we will need to support that API param "forever". Given that and having no explicit use cause for it currently I decided to make that not customizable.

**Review by kraftbj — APPROVED** [review]
This makes sense and works for me. Leaving it as "Needs Review" so another Data Wizard can give it an expert review. Marking it approved so it can be merged after that.

**Review by bperson — APPROVED** [review]
LGTM as well, I'm still not a huge fan of `excess_boost` as a name, but I couldn't come up with anything better than the extra long `fields_boost_multiplier` so 🤷‍♂️

**Comment by kraftbj** [comment]
Noting that the WP.com copy of divergent from Jetpack, so not syncing this at this time.

**Comment by xyu** [comment]
Started a diff to port changes over to WPCOM codebase:

https://code.a8c.com/D19987

**Pull request merged**
